src/fitnesse/wikitext/parser/Headings.java

```diff
@@ -0,0 +1,202 @@
+package fitnesse.wikitext.parser;
+
+import fitnesse.html.HtmlElement;
+import fitnesse.html.HtmlTag;
+import fitnesse.html.HtmlUtil;
+import fitnesse.wiki.BaseWikitextPage;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Stack;
+
+public class Headings extends SymbolType implements Rule, Translation {
+
+  public Headings() {
+    super("Headings");
+    wikiMatcher(new Matcher().startLineOrCell().string("!headings"));
+    wikiRule(this);
+    htmlTranslation(this);
+  }
+
+  @Override
+  public Maybe<Symbol> parse(Symbol current, Parser parser) {
+    Symbol body = parser.parseToEnd(SymbolType.Newline);
+//    for (Symbol option : body.getChildren()) {
+//      if (option.isType(SymbolType.Whitespace)) continue;
+//      if (!option.getContent().startsWith("-")) return Symbol.nothing;
+//      current.add(option);
+//    }
+//    current.evaluateVariables(new String[]{"style"}, parser.getVariableSource());
+    return new Maybe<>(current);
+  }
+
+  @Override
+  public String toTarget(Translator translator, Symbol symbol) {
+    List<HeaderLineParameter> headerLines = new HeaderLineExtracter(translator).getHeaderLines();
+    HeaderLineContentBuilder headerLineContentBuilder = new HeaderLineContentBuilder(headerLines);
+    HtmlElement html = headerLineContentBuilder.html();
+    return html.html();
+  }
+
+  static class HeaderLineExtracter {
+
+    private final Translator translator;
+    private Symbol syntaxTree;
+    private List<Symbol> headerLines = new LinkedList<>();
+    private List<HeaderLineParameter> headerLineParameters = new LinkedList<>();
+
+    private HeaderLineExtracter(final Translator translator) {
+      this.translator = translator;
+    }
+
+    private List<HeaderLineParameter> getHeaderLines() {
+      extractSyntaxTreeFromTranslator();
+      findAllHeaderLinesInSyntaxTree();
+      createHeaderLineParameter();
+      return headerLineParameters;
+    }
+
+    private void findAllHeaderLinesInSyntaxTree() {
+      for (Symbol symbol : syntaxTree.getChildren()) {
+        if (symbol.getType() instanceof HeaderLine) {
+          headerLines.add(symbol);
+        }
+      }
+    }
+
+    private void extractSyntaxTreeFromTranslator() {
+      HtmlTranslator htmlTranslator = (HtmlTranslator) this.translator;
+      ParsingPage parsingPage = htmlTranslator.getParsingPage();
+      WikiSourcePage wikiSourcePage = (WikiSourcePage) parsingPage.getPage();
+      BaseWikitextPage baseWikitextPage = (BaseWikitextPage) wikiSourcePage.getPage();
```

Comment:
There are a couple of problems with this approach. Foremost it's breaking the abstraction between parser and wiki page (the `WikiSourcePage` class). I think it would be better to "ask" the source page for symbols of a particular type, instead of traversing the object graph down to the syntax tree.

On the coding side it would be better to return the SyntaxTree, instead of putting it in an instance variable.